web_poet/serialization/functions.py

```diff
@@ -1,9 +1,55 @@
 import json
-from typing import Type
+from typing import Dict, List, Tuple, Type
+
+from .. import (
+    HttpClient,
+    HttpRequest,
+    HttpRequestBody,
+    HttpRequestHeaders,
+    HttpResponse,
+    HttpResponseBody,
+    HttpResponseHeaders,
+    ResponseUrl,
+)
+from ..page_inputs.url import RequestUrl, _Url
+from .api import (
+    SerializedLeafData,
+    deserialize_leaf,
+    register_serialization,
+    serialize_leaf,
+)
+
+
+def _serialize_HttpRequest(o: HttpRequest) -> SerializedLeafData:
+    other_data = {
+        "url": str(o.url),
+        "method": o.method,
+        "headers": list(o.headers.items()),
+    }
+    result: SerializedLeafData = {
+        "other.json": json.dumps(
+            other_data, ensure_ascii=False, sort_keys=True, indent=2
+        ).encode(),
+    }
+    if o.body:
+        result["body.txt"] = bytes(o.body)
+    return result
 
-from .. import HttpResponse, HttpResponseBody, HttpResponseHeaders, ResponseUrl
-from ..page_inputs.url import _Url
-from .api import SerializedLeafData, register_serialization
+
+def _deserialize_HttpRequest(
+    cls: Type[HttpRequest], data: SerializedLeafData
+) -> HttpRequest:
+    body = HttpRequestBody(data.get("body.txt", b""))
+    other_data = json.loads(data["other.json"])
+    return cls(
+        body=body,
+        url=RequestUrl(other_data["url"]),
+        method=other_data["method"],
+        headers=HttpRequestHeaders(other_data["headers"]),
```

Comment:
I'd expect all these type conversions to be unnecessary, it should be possible to just pass the url and headers. Is it some weird typing issue?